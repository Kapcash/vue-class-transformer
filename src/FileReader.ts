import fs from 'fs';
import ts from 'typescript';
import VueCompiler, { SFCDescriptor } from 'vue-template-compiler'


function isVueFile(filePath: string): boolean {
  return filePath.endsWith('.vue')
} 

/** Recursively get all files from the given folder
 * If the given path is a file, it just returns the file
 * @param path The file or folder path
 * @return All the files present in the folder and its sub-folder, or the file if it's not a folder
 */
function getFilesFromFolder (path = "./") {
  const entries = fs.readdirSync(path, { withFileTypes: true });

  const vueFiles = entries
    .filter(file => !file.isDirectory())
    .map(file => `${path}/${file.name}`)
    .filter(isVueFile);
  const folders = entries.filter(folder => folder.isDirectory());

  for (const folder of folders) {
    vueFiles.push(...getFilesFromFolder(`${path}/${folder.name}`));
  }
  return vueFiles;
}

/** Get all files paths from path and sub-folders */
function getAllFilesPaths (fileOrFolderPath): string[] {
  if (!fs.existsSync(fileOrFolderPath)) throw new Error('Invalid path provided')
  
  if (fs.lstatSync(fileOrFolderPath).isDirectory()) {
    return getFilesFromFolder(fileOrFolderPath)
  } else {
    if (!isVueFile(fileOrFolderPath)) throw new Error('This is not a .vue file!')
    return [fileOrFolderPath]
  };
}

export function getAllFilesToUpgrade (path: string): FileDescriptor[] {
  return getAllFilesPaths(path)
    .map(path => new FileDescriptor(path))
    .filter(file => !!file)
}

export function getSfcDescriptor (file: FileDescriptor): SFCDescriptor | null {
  let sourceStr = fs.readFileSync(file.fullPath, 'utf8')
  
  if (file.isVueFile) {
    return VueCompiler.parseComponent(sourceStr)
  } else {
    return null
  }
}

export function extractScriptFromSfc(sfc: SFCDescriptor): ts.SourceFile {
  return ts.createSourceFile('inline.ts', sfc.script.content, ts.ScriptTarget.ES2020)
}

/** A class descripting a file main attributes */
export class FileDescriptor {
  /** @see https://regex101.com/r/skRAGV/1 */
  private MATCH_SCRIPT_FILE_PATH = /^(.*\/)?(.+)\.(vue|ts|js)$/

  fullPath: string = ''
  path: string = ''
  name: string = ''
  extension: string = ''
  
  constructor(fullPath: string) {
    const [ _fullPath, path, name, extension ] = fullPath.match(this.MATCH_SCRIPT_FILE_PATH)

    if (!_fullPath) return null

    this.fullPath = fullPath
    this.path = path
    this.name = name
    this.extension = extension
  }

  get isVueFile(): boolean {
    return this.extension === 'vue'
  }

  get nameWithExtension(): string {
    return `${this.name}.${this.extension}`
  }
}
