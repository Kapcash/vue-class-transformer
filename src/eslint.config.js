module.exports = {
  "space-before-blocks": "warn",
  "padding-line-between-statements": [
    "error",
    { blankLine: "always", prev: "*", next: "*"},
  ],
  "@typescript-eslint/member-delimiter-style": ["error"],
  semi: "off",
  "@typescript-eslint/semi": ["error", "always"],
  "no-return-assign": ["error", "except-parens"],
  "comma-dangle": ["error", "always-multiline"],
  "lines-between-class-members": ["error", "always", { exceptAfterSingleLine: true }],
  "quote-props": ["error", "as-needed"],
  "space-before-function-paren": ["error", {
    anonymous: "never",
    named: "never",
    asyncArrow: "always",
  }],
  "import/no-mutable-exports": "off",
  "no-unused-vars": "off",
  "import/order": "off",
  "no-console": "off",
  camelcase: "off",
}
