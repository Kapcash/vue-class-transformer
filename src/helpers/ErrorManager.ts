export class ErrorManager {
  errorMap = new Map<string, Error[]>();

  addError(fileName: string, message: Error) {
    if (this.errorMap.has(fileName)) {
      const errorEntry = this.errorMap.get(fileName);
      errorEntry.push(message);
    } else {
      this.errorMap.set(fileName, [message]);
    }
  }

  toString() {
    const errorLogs: string[] = [];
    for(const [name, errors] of this.errorMap.entries()) {
      errorLogs.push(`- ${name}: ${errors.map(err => err.stack || err.message)}`);
    }
    return errorLogs.join('\n\n');
  }
}