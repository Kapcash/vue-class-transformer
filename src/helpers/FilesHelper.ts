import fs from 'fs';
import { dirname } from "path";
import mkdirp from 'mkdirp'
import ts from 'typescript';
import * as VueCompiler from 'vue-template-compiler'
import { SFCDescriptor, SFCBlock } from 'vue-template-compiler'

export function getAllFilesToUpgrade(inputPaths: string[]): FileDescriptor[] {
  const vueFiles = inputPaths
    .map(path => new FileDescriptor(path))
    .filter(file => file.isVueFile);
  if (vueFiles.length <= 0) { throw new Error("This path doesn't include any .vue file!"); }
  return vueFiles;
}

/** Extract the Vue SFC descriptor from an input file */
export function getSfcDescriptor (file: FileDescriptor): SFCDescriptor | null {
  const sourceStr = fs.readFileSync(file.fullPath, 'utf8')
  
  if (file.isVueFile) {
    return VueCompiler.parseComponent(sourceStr)
  } else {
    return null
  }
}

/** Extract the component script as a typescript AST */
export function extractScriptFromSfc(vueFile: FileDescriptor) {
  const sfc = getSfcDescriptor(vueFile)
  let { content, src, start, end } = sfc.script
  const tsScriptPath = `${vueFile.path}${src}`

  if (src && !content) {
    content = fs.readFileSync(`${process.cwd()}/${tsScriptPath}`).toString()
    start = end = NaN
  }

  const sourceScript = ts.createSourceFile('inline.ts', content, ts.ScriptTarget.ES2020)

  return {
    sourceScript,
    tsScriptPath,
    start,
    end,
  }
}

/**
 * Replace the Vue SFC script by another. Can create a new duplicated file instead of replacing the existing input.
 * @param filePath The input file path to replace the script
 * @param outputPath The output file path to write the new file. May be equal to the input path.
 * @param scriptBlock The Vue SFC block from the input file
 * @param scriptString The new script to write to the output file / replace existing script
 */
export function replaceVueScript(filePath: string, outputPath: string, scriptString: string, scriptStart: number = 0, scriptEnd: number = 0) {
  scriptStart ||= 0
  scriptEnd ||= 0

  mkdirp.sync(dirname(outputPath))
  const fd = fs.openSync(filePath, 'r')

  const stats = fs.fstatSync(fd);

  const beforeSize = scriptStart
  const beforeBuffer = Buffer.alloc(beforeSize)
  if (beforeSize) {
    fs.readSync(fd, beforeBuffer, null, beforeSize, null)
  }

  const afterSize = stats.size - scriptEnd
  const afterBuffer = Buffer.alloc(afterSize)
  fs.readSync(fd, afterBuffer, null, afterSize, scriptEnd)

  const scriptBuffer = Buffer.from(scriptString, 'utf8')

  fs.writeFileSync(outputPath, Buffer.concat([beforeBuffer, scriptBuffer, afterBuffer]), { encoding: 'utf-8' })
}

/** A class descripting a file main attributes */
export class FileDescriptor {
  /** @see https://regex101.com/r/skRAGV/1 */
  private MATCH_SCRIPT_FILE_PATH = /^(.*\/)?(.+)\.(vue|ts|js)$/

  fullPath: string = ''
  path: string = ''
  name: string = ''
  extension: string = ''
  
  constructor(fullPath: string) {
    const [ _fullPath, path, name, extension ] = fullPath.match(this.MATCH_SCRIPT_FILE_PATH) || []

    if (!_fullPath) return null

    this.fullPath = fullPath
    this.path = path
    this.name = name
    this.extension = extension
  }

  get isVueFile(): boolean {
    return this.extension === 'vue'
  }

  get nameWithExtension(): string {
    return `${this.name}.${this.extension}`
  }
}
