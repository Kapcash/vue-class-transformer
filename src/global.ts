import { RuntimeConfiguration } from './helpers/ArgumentParser';
import { ErrorManager } from './helpers/ErrorManager';

export type SFCToken = 'script' | 'template' | 'styles' | 'other';
export type AttributeToken = 'data' | 'props' | 'computed' | 'watcher' | 'hooks' | 'methods' | 'other';

export function Alias(parameter: AttributeToken | SFCToken) {
  return function (target, propertyKey: string, descriptor: PropertyDescriptor) {
    target[parameter] = descriptor.value;
  };
}

Array.prototype.chunk = function(statement) {
  return this.reduce(
    ([validChunk, invalidChunk], value) => {
      const targetChunk = statement(value) ? validChunk : invalidChunk;
      targetChunk.push(value);
      return [validChunk, invalidChunk];
    },
    [[], []]
  );
};

declare global {
  namespace NodeJS {
    interface Global {
      config: RuntimeConfiguration;
      errors: ErrorManager;
    }
  }

  interface Array<T> {
    chunk(statement: (val: T) => boolean): [Array<T>, Array<T>];
  }
}
