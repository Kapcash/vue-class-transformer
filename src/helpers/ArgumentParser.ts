import minimist from 'minimist'
import chalk from 'chalk';
import { AttributeToken, SFCToken } from '../global.js';

export interface Configuration {
  overrideFiles: boolean;
  isNuxt: boolean;
  outputDir: string;
  sfcOrder: SFCToken[];
  propertiesOrder: AttributeToken[];
}

export interface RuntimeConfiguration extends Configuration {
  inputPath: string;
}

export async function parseArguments(args): Promise<RuntimeConfiguration> {
  // Arguments parsing
  const argv = minimist(args, {
    boolean: ['n', 'f', 'h'],
    alias: {
      n: 'nuxt',
      h: 'help',
      f: 'force',
      c: 'config',
      o: 'output',
    },
    default: {
      n: false,
      f: false,
      o: './generated',
    }
  })

  const { _: argPath, config: configPath, help, nuxt, force, output } = argv
  
  if (help) {
    return Promise.reject(false)
  }
    
  let options: Configuration = {
    isNuxt: nuxt,
    overrideFiles: force,
    outputDir: output,
    sfcOrder: ['script', 'template', 'styles', 'other'],
    propertiesOrder: ['data', 'props', 'watcher', 'hooks', 'methods', 'computed', 'other'],
  }

  if (configPath) {
    const configAbsPath = `${process.cwd()}/${configPath}`
    try {
      const configFile: Configuration = (await import(configAbsPath)).default
      options = { ...options, ...configFile }
    } catch (err) {
      console.error(`Can't find the configuration file at ${configAbsPath}.`, err)
    }
  }
  
  const inputPath = argPath[0]
  const outputDir = options.outputDir.endsWith('/') ? options.outputDir : `${options.outputDir}/`

  const runtimeConfig: RuntimeConfiguration = {
    ...options,
    outputDir,
    inputPath,
  }
  
  return runtimeConfig;
}

export function printHelp() {
  const info = console.log;
  info(`${chalk.bold('Syntax:')}   vts [options] [file|folder path]`)
  
  info('')

  info(`Example:  vts -n -f ./src/test.vue`)
  info(`Example:  vts -o dist/ ./src/components/`)
  
  info('')

  info(chalk.bold(`Options:`))
  info(` -n, --nuxt          Import decorators from 'nuxt-property-decorator' library. Default is 'vue-property-decorator'.`)
  info(` -f, --force         Replace converted files in place instead of created new files.`)
  info(` -o, --output        Path to the output folder. Ignored if option --force is used.`)
  info(` -c, --config         Path to a configuration file. The configuration file options override CLI options if they collide.`)
  info(` -h, --help          Print this helper.`)
  info(` -v, --version       Print the CLI current version.`)
  
  info('')

  info(`See the github README.md for more information: https://github.com/Kapcash/vue-typescript-converter#readme`)
}