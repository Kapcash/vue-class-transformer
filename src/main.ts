import ts from 'typescript';
import { ESLint, Linter } from 'eslint';
import { sourceExtractor } from './extractors/NodeExtractor';
import VueComponentDescriptor from './extractors/VueComponentDescriptor';
import { FileDescriptor, replaceVueScript, extractScriptFromSfc } from './helpers/FilesHelper';
import VuePropertyDecoratorBuilder from './builders/component/ComponentBuilders';
import ComponentDirector from './builders/component/ComponentDirector';
import eslintBaseConfig from './eslint.config';

/** Converts a Vue SFC script to a class based syntax */
export async function upgradeComponent(vueFile: FileDescriptor) {
  const { sourceScript, tsScriptDescriptor, start, end } = convertScript(vueFile);

  const scriptFile = isNaN(start) ? tsScriptDescriptor : vueFile;
  const outputPath = global.config.overrideFiles ? scriptFile.fullPath : `${global.config.outputDir}${scriptFile.nameWithExtension}`;

  const prettyScript = await lintScript(sourceScript);

  replaceVueScript(scriptFile.fullPath, outputPath, prettyScript || sourceScript, start, end);
}

/** Convert the given SFC descriptor's script */
export function convertScript(vueFile: FileDescriptor) {
  const { sourceScript, tsScriptPath, start, end } = extractScriptFromSfc(vueFile);
  
  const vueDescriptor = new VueComponentDescriptor();
  // For every first level child in the source, we extract all the info we need
  ts.forEachChild(sourceScript, sourceExtractor(vueDescriptor));
  
  const componentBuilder = new VuePropertyDecoratorBuilder(global.config.isNuxt);
  componentBuilder.setDescriptor(vueDescriptor);
  const director = new ComponentDirector(componentBuilder, global.config.propertiesOrder);
  const newScript = director.build();

  return {
    sourceScript: ts.createPrinter().printFile(newScript),
    tsScriptDescriptor: new FileDescriptor(tsScriptPath),
    start,
    end,
  };
}

async function lintScript(sourceScript: string): Promise<string> {
  // FIXME: Webpack loading the parser doesn't work on production bundle (including all compiled node_modules, not as external)
  const baseConfig: Linter.Config = {
    parser: '@typescript-eslint/parser',
    plugins: [
      '@typescript-eslint',
    ],
    rules: eslintBaseConfig,
  };
  
  const eslint = new ESLint({
    fix: true,
    useEslintrc: false,
    baseConfig,
  });

  const lintedScript = await eslint.lintText(sourceScript);

  return lintedScript[0].output;
}