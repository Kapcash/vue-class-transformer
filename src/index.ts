import { getAllFilesToUpgrade, FileDescriptor } from './helpers/FileReader';
import { parseArguments } from './helpers/ArgumentParser';
import { upgradeComponent } from './main'
import cliProgress from 'cli-progress'

// ===== MAIN ===== //

(async function() {
  global.config = parseArguments()

  const progressBar = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
  
  try {
    const vueFilesToUpgrade: FileDescriptor[] = getAllFilesToUpgrade(global.config.inputPaths)
    progressBar.start(vueFilesToUpgrade.length, 0)
    
    const upgradeVueComponent = upgradeComponent(global.config)
    vueFilesToUpgrade.forEach((file) => {
      upgradeVueComponent(file)
      progressBar.increment();
    })
  } catch (err) {
    console.error(err.message)
  } finally {
    progressBar.stop()
  }
})()

// ===================== //
