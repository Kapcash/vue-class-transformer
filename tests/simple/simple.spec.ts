import { convertScript } from '../../src/main';
import { FileDescriptor } from '../../src/helpers/FilesHelper';

describe('Simple conversions', () => {
  global.config = {
    inputPaths: [],
    outputDir: '',
    overrideFiles: false,
    isNuxt: false,
    propertiesOrder: ['data', 'props', 'watcher', 'hooks', 'methods', 'computed', 'other'],
  };

  it('Convert an empty Vue SFC', () => {
    const testPath = __dirname + '/empty.vue';
    const convertedScript = convertScript(new FileDescriptor(testPath));
    expect(convertedScript.sourceScript).toMatchSnapshot();
  });

  it('Convert a simple Vue SFC', () => {
    const testPath = __dirname + '/simple.vue';
    const convertedScript = convertScript(new FileDescriptor(testPath));
    expect(convertedScript.sourceScript).toMatchSnapshot();
  });

  it('Do not remove eventual custom options in SFC', () => {
    const testPath = __dirname + '/customOption.vue';
    const convertedScript = convertScript(new FileDescriptor(testPath));
    expect(convertedScript.sourceScript).toMatchSnapshot();
  });
  it('Transform both getters and setters in computed', () => {
    const testPath = __dirname + '/getterSetter.vue';
    const convertedScript = convertScript(new FileDescriptor(testPath));
    expect(convertedScript.sourceScript).toMatchSnapshot();
  });
});
