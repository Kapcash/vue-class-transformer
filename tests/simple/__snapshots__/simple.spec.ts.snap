// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple conversions Convert a simple Vue SFC 1`] = `
"import { Vue, Component, Watch, Prop } from 'vue-property-decorator';
import { TextField } from \\"~/components/test-field.vue\\";
@Component<MyComponent>({ components: { TextField } })
export default class MyComponent extends Vue {
    loading = true;
    isOpen = false;
    query = \\"\\";
    @Prop({
        type: Boolean, required: false, default: false
    })
    readonly loading!: Boolean;
    @Prop({
        type: String, required: true
    })
    readonly userId!: String;
    @Watch('loading')
    onLoadingUpdate(newVal, oldVal) {
        console.log(newVal);
    }
    close() {
        this.isOpen = false;
    }
    get hasQuery() {
        return !!this.query;
    }
    created() {
        this.watcher = this.$store.watch(() => this.$store.state.modals.isResetPasswordActive, (newVal) => {
            this.isOpen = newVal;
        });
    }
}
"
`;

exports[`Simple conversions Convert an empty Vue SFC 1`] = `
"import { Vue, Component } from 'vue-property-decorator';
import { TextField } from \\"~/components/test-field.vue\\";
@Component
export default class EmptyComponent extends Vue {
}
"
`;

exports[`Simple conversions Do not remove eventual custom options in SFC 1`] = `
"import { Vue, Component, Watch, Prop } from 'vue-property-decorator';
import { TextField } from \\"~/components/test-field.vue\\";
@Component<MyComponent>({ components: { TextField }, i18n: {
        lang: \\"en\\"
    } })
export default class MyComponent extends Vue {
    loading = true;
    isOpen = false;
    query = \\"\\";
    @Prop({
        type: Boolean, required: false, default: false
    })
    readonly loading!: Boolean;
    @Prop({
        type: String, required: true
    })
    readonly userId!: String;
    @Watch('loading')
    onLoadingUpdate(newVal, oldVal) {
        console.log(newVal);
    }
    close() {
        this.isOpen = false;
    }
    get hasQuery(): boolean {
        return !!this.query;
    }
    created() {
        this.watcher = this.$store.watch(() => this.$store.state.modals.isResetPasswordActive, (newVal) => {
            this.isOpen = newVal;
        });
    }
}
"
`;

exports[`Simple conversions Transform both getters and setters in computed 1`] = `
"import { Vue, Component } from 'vue-property-decorator';
@Component
export default class  extends Vue {
    get currentValue(): string {
        return this.$store.state.title;
    }
    set currentValue(val: string) {
        this.$store.commit(\\"SET_TITLE\\", val);
    }
}
"
`;
